src/model_theory/terms_and_formulas.lean

```diff
@@ -253,5 +267,91 @@ begin
   rw [fin.coe_nat_eq_last, fin.snoc_last],
 end
 
+namespace Theory
+
+/-- A theory is satisfiable if a structure models it. -/
+def is_satisfiable (T : L.Theory) : Prop :=
+∃ (M : Type (max u v)) [nonempty M] [str : L.Structure M], @Theory.model L M str T
+
+/-- A theory is finitely satisfiable if all of its finite subtheories are satisfiable. -/
+def is_finitely_satisfiable (T : L.Theory) : Prop :=
+∀ (T0 : finset L.sentence), (T0 : L.Theory) ⊆ T → (T0 : L.Theory).is_satisfiable
+
+/-- Given that a theory is satisfiable, selects a model using choice. -/
+def is_satisfiable.some_model {T : L.Theory} (h : T.is_satisfiable) : Type* :=
+  classical.some h
+
+instance is_satisfiable.nonempty_some_model {T : L.Theory} (h : T.is_satisfiable) :
+  nonempty (h.some_model) :=
+classical.some (classical.some_spec h)
+
+noncomputable instance is_satisfiable.inhabited_some_model {T : L.Theory} (h : T.is_satisfiable) :
+  inhabited (h.some_model) := classical.inhabited_of_nonempty'
+
+noncomputable instance is_satisfiable.some_model_structure {T : L.Theory} (h : T.is_satisfiable) :
+  L.Structure (h.some_model) :=
+classical.some (classical.some_spec (classical.some_spec h))
+
+lemma is_satisfiable.some_model_realize_theory {T : L.Theory} (h : T.is_satisfiable) :
+  T.model h.some_model :=
+classical.some_spec (classical.some_spec (classical.some_spec h))
+
+lemma is_satisfiable.of_model {T : L.Theory} (M : Type (max u v)) [n : nonempty M]
+  [S : L.Structure M] (h : T.model M) : T.is_satisfiable :=
+⟨M, n, S, h⟩
+
+lemma is_satisfiable.is_satisfiable_subset {T T' : L.Theory} (h : T'.is_satisfiable) (hs : T ⊆ T') :
+  T.is_satisfiable :=
+⟨h.some_model, h.nonempty_some_model, h.some_model_structure,
+  realize_theory_subset h.some_model_realize_theory hs⟩
+
+lemma is_satisfiable.is_finitely_satisfiable {T : L.Theory} (h : T.is_satisfiable) :
+  T.is_finitely_satisfiable :=
+λ _, h.is_satisfiable_subset
+
+variable {n : ℕ}
+
+/-- Two (bounded) formulas are semantically equivalent over a theory `T` when they have the same
+interpretation in every model of `T`. (This is also known as logical equivalence, which also has a
+proof-theoretic definition.) -/
+def semantically_equivalent (T : L.Theory) (φ ψ : L.bounded_formula α n) : Prop :=
+∀ (M : Type (max u v)) (str : L.Structure M), @model L M str T →
+  @realize_bounded_formula L M str _ _ φ = @realize_bounded_formula L M str _ _ ψ
+
+lemma semantically_equivalent_model {T : L.Theory} {φ ψ : L.bounded_formula α n}
```

Comment:
Maybe allow projection notation for this and the following lemmas. Not sure what to suggest here. `semantically_equivalent.realize_eq`?